community/planning/consensus_runner.md

```diff
@@ -0,0 +1,306 @@
+        String,
+        Box<
+            dyn Algorithm<
+                Event = ConsensusEvent,
+                Action = ConsensusAction,
+                Context = ConsensusContext,
+            >,
+        >,
+    >,
+    consensus_store_command_factory:
+        ConsensusStoreCommandFactory<<E as StoreCommandExecutor>::Context>,
+    store_command_executor: Arc<E>,
+}
+```
+
+The `ConsensusRunner` effective runs the following loop:
+
+```rust
+fn run(&self, service_id: &FullyQualifiedServiceId)
+    -> Result<(), InternalError>
+{
+    let store = self.pooled_scabbard_store_factory.new_store();
+
+    let unprocessed_actions = store.list_consensus_actions(service_id)?;
+
+    if !unprocessed_actions.is_empty() {
+        // run each action and execute the commands before running the next action
+        for action in unprocessed_actions {
+            let commands = self.action_runner.run_actions(vec![action], service_id)?;
+            self.store_command_executor.execute(commands)?;
+        }
+    }
+
+    let unprocessed_event = store
+        .list_consensus_events(service_id)?
+        .get(0)
+        .cloned();
+
+    let event = match unprocessed_event {
+        Some(event) => event,
+        None => {
+            // No events
+            return Ok(());
+        }
+    };
+
+    let (event_id, event) = event.deconstruct();
+
+    let context = store
+        .get_current_consensus_context(service_id)?
+        .ok_or_else(|| {
+            InternalError::with_message(format!(
+                "No scabbard context for service {}",
+                service_id
+            ))
+        })?;
+
+    let epoch = context.epoch();
+
+    let algorithm = self.algorithms.get(event.algorithm_name()).ok_or_else(|| {
+        InternalError::with_message(format!("{} is not configured", event.algorithm_name()))
+    })?;
+    let actions = algorithm
+        .event(event, context)
+        .map_err(|e| InternalError::from_source(Box::new(e)))?;
+
+    let commands = vec![
+        self.consensus_store_command_factory
+            .new_save_actions_command(service_id, actions, event_id),
+        self.consensus_store_command_factory
+            .new_mark_event_complete_command(service_id, event_id, epoch),
+    ];
+    self.store_command_executor.execute(commands)?;
+
+    // run the resulting actions
+    let unprocessed_actions = store.list_consensus_actions(service_id)?;
+
+    if !unprocessed_actions.is_empty() {
+        // run each action and execute the commands before running the next action
+        for action in unprocessed_actions {
+            let commands = self.action_runner.run_actions(vec![action], service_id)?;
+            self.store_command_executor.execute(commands)?;
+        }
+    }
+
+    Ok(())
+}
+```
+
+This run function references several traits: `UnprocessedActionSource`, and
+`UnprocessedEventSource`.
+
+`UnprocessedActionSource` is fairly simple:
+
+```rust
+trait UnprocessedActionSource {
```

Comment:
Some of these traits have been removed/are not used. They should also be removed from this doc.

Reply:
Removed, and updated the other traits that exist as structs.